### ISSUE TITLE
Interface Desing

### ISSUE THREAD

**Issue by Thomas-P** [alexa/alexa-skills-kit-sdk-for-nodejs#235]
Hi, 

I am not sure, but is it a good idea to provide an interface as a context object? The main problem is, that you have to deal with so many side affects. Think about the this.t.
```javacript
function test() {
  this.t('something');
}
```
If I want a maintainable function, that shows me what **this.t** is, I can do the following.
```javacript
function test() {
  const translate = this.t;
  translate('something');
}
```
This wil cause an error, because we are now out of the context. I have to bind it manually, like this:
```javacript
function test() {
  const translate = this.t.bind(this);
  translate('something');
}
```
But this is not an ideal programmer friendly sollution. In a unit test, where I test if translate is called, I won't get an error. The error will throw in the lambda console, where I don't get a proper message, about what happened. There is a index out of bound error or something similiar, that won't help. 

Another problem are fat arrow functions. If I choose to implement my skill by using theese, I could not access the skill kit interface, because the in fat arrow function, the context can't be overwritten.

And another problem is, that using an api as a context, I don't have an overview about what is happened under the hood. Any function that is called did not have a defined interface or prototype functions, that I could check if something happened. It was for me a really bad experience, because I couldn't check the code, if something unexpected happened. 

A way around this could be to provide the api as a parameter in the intent function. This will avoid the problems with the context. On the start of an intent callback, we have the state object as an parameter and we could use them as expected. On the other hand is to mention, that this sollution will break earlier versions. 

Another way is to bind all functions, that are provided in the interface to the context object. This could help to avoid context losses, when we use variables to store information, but will not work, when we use arrow functions for the intent object.

In summary, we see, that if we are using an api as a context object, we have to deal with side effects, based on the language specification. To avoid this, we should use an api, that is given by an parameter. This will help to create a good skill experience and avoid errors which are hard to predict by simple testing. But this also means, that we have to deal with a breaking api.

**Comment by tianrenz** [comment]
Hi @Thomas-P, 

Thanks for your thoughtful input. We are currently in the process of redesigning the sdk. Your ideas are certainly very illuminating. We will keep them in mind and try to solve those complications in the new design. Stay tuned! 

Regards
